crates/trie/trie/src/updates.rs

```diff
@@ -36,14 +36,32 @@ impl TrieUpdates {
 
     /// Extends the trie updates.
     pub fn extend(&mut self, other: Self) {
-        self.account_nodes.retain(|nibbles, _| !other.removed_nodes.contains(nibbles));
-        self.account_nodes.extend(ExcludeEmptyFromPair::from_iter(other.account_nodes));
-        self.removed_nodes.extend(ExcludeEmpty::from_iter(other.removed_nodes));
+        self.extend_common(&other);
+        self.account_nodes.extend(exclude_empty_from_pair(other.account_nodes));
+        self.removed_nodes.extend(exclude_empty(other.removed_nodes));
         for (hashed_address, storage_trie) in other.storage_tries {
             self.storage_tries.entry(hashed_address).or_default().extend(storage_trie);
         }
     }
 
+    /// Extends the trie updates.
+    ///
+    /// Slightly less efficient than [`Self::extend`], but preferred to `extend(other.clone())`.
+    pub fn extend_ref(&mut self, other: &Self) {
+        self.extend_common(other);
+        self.account_nodes.extend(exclude_empty_from_pair(
+            other.account_nodes.iter().map(|(k, v)| (k.clone(), v.clone())),
+        ));
+        self.removed_nodes.extend(exclude_empty(other.removed_nodes.iter().cloned()));
```

Comment:
this can be moved into extend common too

Reply:
or we can just get rid of `extend_common`, unsure why we need this as a separate method

Reply:
cloned vs into_iter